### ISSUE TITLE
Add a logical clock to the validators' windows

### ISSUE THREAD

**Issue by garious** [solana-labs/solana#309]
Validators currently can't figure out what entries its missing. Maintain an `entry_count` in `WriteStage` with a meaning analogous to Bitcoin's block height, and add that to `Blob` metadata. Use it to order blobs and identify missing entries.

**Comment by garious** [comment]
Looks like `ledger::to_blobs()` might need to be extended to take an `mut entry_count` parameter and that it be used in the `Blob` metadata. If that data were there, we could scramble the order of the returned `VecDeque` and still be able to reconstruct it or identify missing blobs.

It's not clear to me what else needs to be included or why.

**Comment by garious** [comment]
Let's decide, should blobs be ordered? Request/Response doesn't need that feature. Seems like we need a torrent transport layer with similiar high-level properties of TCP.

**Comment by rob-solana** [comment]
The approach I'm going to try to implement is to basically whack Block::to_blobs() in favor of Entry::to_blobs().  The change implies that there will be at most one entry per blob.

This change should allow a node starting from a transaction log (which consists ony of Entries) to infer a blob index for the network up to that point, and thereby be able to synchronize with a running network (which would have the same mapping of entry<=>blob index.

This solution obviates the need for an Entry index.

**Comment by garious** [comment]
Hoping that `Block::to_blobs()` is reduced to:

```rust
self.into_iter().flat_map(|entry| entry.to_blobs()).collect()
```

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map

**Comment by garious** [comment]
@rob-solana, any more work happening under this ticket?

**Comment by rob-solana** [comment]
I've got the bank counting Entries, with which I think I can initialize windows.

I'm worried this isn't a complete enough solution.   I also need to write a test that fails.

**Comment by garious** [comment]
Are you able to PR the initialized window from entry count? If that's incomplete, would the additional work fall under #299?

**Comment by rob-solana** [comment]
PR coming right now.

need help with the test, though

**Comment by rob-solana** [comment]
@mvines did reverting fix you?

**Comment by mvines** [comment]
No, I was suffering from #446

**Comment by rob-solana** [comment]
with PR #467 the network should have a unified "Entry height" that manifests as a blob index

**Comment by garious** [comment]
Fixed in #467
